CMakeLists.txt

```diff
@@ -0,0 +1,345 @@
+		if(NOT ${RESOURCE_NAME} MATCHES "SETTING")
+			file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}/${file_name})
+
+			set(OUTPUT_OBJECT ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}/${file_name}.o)
+			set(OUTPUT_OBJECT_D ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}/${file_name}.d)
+
+			# This command generates the object file and dependency? of the exported ASM and header
+			# OUTPUT again means these files are set as GENERATED
+			# MAIN_DEPENDENCY sets the input ASM file as the main dependency
+			# DEPENDS sets the header file as a dependency
+			# We need too set dependencies to the input files used in the command, as it is not automatically done even though they are source files.
+			# This ensures this command is ran after generating the ASM and header files
+			add_custom_command(OUTPUT ${OUTPUT_OBJECT} ${OUTPUT_OBJECT_D}
+				MAIN_DEPENDENCY ${OUTPUT_FILE_1}
+				DEPENDS ${OUTPUT_FILE_2}
+				COMMAND ${GBA_C_COMPILER} -MMD -MP -MF ${OUTPUT_OBJECT_D} -x assembler-with-cpp ${ASFLAGS} -c ${OUTPUT_FILE_1} -o ${OUTPUT_OBJECT}
+				COMMENT "BUILDING ${OUTPUT_OBJECT} AND ${OUTPUT_OBJECT_D}")
+
+			# Since we pass in dependencies, we can handle all resource objects as a group
+			# We append to a lists that we later set to the parent scope to use when building
+			list(APPEND RESOURCE_OBJECT_FILES ${OUTPUT_OBJECT})
+			list(APPEND RESOURCE_OBJECT_D_FILES ${OUTPUT_OBJECT_D})
+		endif()
+
+		# The second output file is usually the header file,
+		# unless the settings are being made, then it is the first.
+		if(NOT ${RESOURCE_NAME} MATCHES "SETTING")
+			list(APPEND RESOURCES_HPP ${OUTPUT_FILE_2})
+		else()
+			list(APPEND RESOURCES_HPP ${OUTPUT_FILE_1})
+		endif()
+	endforeach()
+
+	# Set our object, dependency? and header lists to the parent scope
+	set(RESOURCE_OBJECT_FILES ${RESOURCE_OBJECT_FILES} PARENT_SCOPE)
+	set(RESOURCE_OBJECT_D_FILES ${RESOURCE_OBJECT_D_FILES} PARENT_SCOPE)
+	set(RESOURCES_HPP ${RESOURCES_HPP} PARENT_SCOPE)
+endfunction()
+
+function(build_game_objects)
+	# The files we want to build
+	set(GAME_SOURCE_FILES ${ARGV0})
+
+	# The name we use to create the lists of output object files, it should be unique and descriptive
+	set(SOURCE_TYPE ${ARGV1})
+	foreach (FILE ${GAME_SOURCE_FILES})
+		get_filename_component(file_dir "${CMAKE_BINARY_DIR}/source/${FILE}" PATH)
+		string(REPLACE "/" ";" splits ${FILE})
+		list(GET splits -1 file_new)
+		string(REPLACE "." ";." file_new_split ${file_new})
+		list(GET file_new_split 0 file_name)
+		list(GET file_new_split 1 file_type)
+		file(MAKE_DIRECTORY ${file_dir}/${file_name})
+		if(file_type MATCHES ".iwram")
+			list(GET file_new_split 2 file_type2)
+			set(OUTPUT_OBJECT ${file_dir}/${file_name}/${file_name}.iwram.o)
+			set(OUTPUT_OBJECT_D ${file_dir}/${file_name}/${file_name}.iwram.d)
+		else()
+			set(OUTPUT_OBJECT ${file_dir}/${file_name}/${file_name}.o)
+			set(OUTPUT_OBJECT_D ${file_dir}/${file_name}/${file_name}.d)
+		endif()
+		
```

Comment:
Oh, now I understand you are using custom commands for everything here! We should discuss a bit more, because the `.d` files generated are important for dependency checking for the makefile build. Since CMake has no built-in way of automatically checking for dependencies (since it relies on the target build system to do so), we would have to think of another solution here. Pregenerating the dependencies, maybe?

Reply:
So, I use 2 custom targets, and the GENERATED feature of cmake. Command outputs are automatically set as generated, so they are included in the cleans too. The 2 targets just have the objects to be generated set as dependencies, and the first set of resources generated, the sprites, are dependent on the tools target, which is an external project that builds the export tool. GLOB has an option in it, CONFIGURE DEPENDS that means it rechecks when you build. Sorry if I'm misunderstanding what you're trying to tell me.